File: backend/routers/config_sync/config_sync.py
"""Config-Sync Service Router.

API endpoints for managing VyOS primary/secondary configuration synchronization.
The template structure is identical between VyOS 1.4 and 1.5.

Endpoints:
  GET  /vyos/config-sync/capabilities  — version-aware feature flags
  GET  /vyos/config-sync/config        — normalized config-sync configuration
  POST /vyos/config-sync/batch         — atomic set/delete operations
"""

from fastapi import APIRouter, HTTPException, Request
from starlette.concurrency import run_in_threadpool
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from session_vyos_service import get_session_vyos_service
from vyos_builders.config_sync import ConfigSyncBatchBuilder
from fastapi_permissions import require_read_permission, require_write_permission
from rbac_permissions import FeatureGroup
import inspect
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/vyos/config-sync", tags=["config-sync"])


# ============================================================================
# Pydantic Models
# ============================================================================


class ConfigSyncSecondary(BaseModel):
    address: Optional[str] = None
    key: Optional[str] = None
    port: Optional[int] = None
    timeout: Optional[int] = None


class ConfigSyncSections(BaseModel):
    # Simple top-level presence flags
    firewall: bool = False
    nat: bool = False
    nat66: bool = False
    pki: bool = False
    policy: bool = False
    vpn: bool = False
    vrf: bool = False

    # Interfaces (parent + subtypes)
    interfaces: bool = False
    interfaces_bonding: bool = False
    interfaces_bridge: bool = False
    interfaces_dummy: bool = False
    interfaces_ethernet: bool = False
    interfaces_geneve: bool = False
    interfaces_input: bool = False
    interfaces_l2tpv3: bool = False
    interfaces_loopback: bool = False
    interfaces_macsec: bool = False
    interfaces_openvpn: bool = False
    interfaces_pppoe: bool = False
    interfaces_pseudo_ethernet: bool = False
    interfaces_sstpc: bool = False
    interfaces_tunnel: bool = False
    interfaces_virtual_ethernet: bool = False
    interfaces_vti: bool = False
    interfaces_vxlan: bool = False
    interfaces_wireguard: bool = False
    interfaces_wireless: bool = False
    interfaces_wwan: bool = False

    # Protocols (parent + subtypes)
    protocols: bool = False
    protocols_babel: bool = False
    protocols_bfd: bool = False
    protocols_bgp: bool = False
    protocols_failover: bool = False
    protocols_igmp_proxy: bool = False
    protocols_isis: bool = False
    protocols_mpls: bool = False
    protocols_nhrp: bool = False
    protocols_ospf: bool = False
    protocols_ospfv3: bool = False
    protocols_pim: bool = False
    protocols_pim6: bool = False
    protocols_rip: bool = False
    protocols_ripng: bool = False
    protocols_rpki: bool = False
    protocols_segment_routing: bool = False
    protocols_static: bool = False

    # QoS (parent + subtypes)
    qos: bool = False
    qos_interface: bool = False
    qos_policy: bool = False

    # Service (parent + subtypes)
    service: bool = False
    service_console_server: bool = False
    service_dhcp_relay: bool = False
    service_dhcp_server: bool = False
    service_dhcpv6_relay: bool = False
    service_dhcpv6_server: bool = False
    service_dns: bool = False
    service_lldp: bool = False
    service_mdns: bool = False
    service_monitoring: bool = False
    service_ndp_proxy: bool = False
    service_ntp: bool = False
    service_snmp: bool = False
    service_tftp_server: bool = False
    service_webproxy: bool = False

    # System (parent + subtypes)
    system: bool = False
    system_conntrack: bool = False
    system_flow_accounting: bool = False
    system_login: bool = False
    system_option: bool = False
    system_sflow: bool = False
    system_static_host_mapping: bool = False
    system_sysctl: bool = False
    system_time_zone: bool = False


class ConfigSyncConfig(BaseModel):
    mode: Optional[str] = None
    secondary: Optional[ConfigSyncSecondary] = None
    sections: ConfigSyncSections = Field(default_factory=ConfigSyncSections)


class ConfigSyncBatchOperation(BaseModel):
    op: str = Field(..., description="Builder method name")
    value: Optional[str] = Field(
        None,
        description=(
            "Argument(s) for the operation. "
            "Single-arg methods: plain string. "
            "Multi-arg methods: comma-separated (e.g., 'interfaces,ethernet')."
        ),
    )


class ConfigSyncBatchRequest(BaseModel):
    operations: List[ConfigSyncBatchOperation]


class VyOSResponse(BaseModel):
    success: bool
    data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None


# ============================================================================
# Internal builder method denylist
# ============================================================================

_INTERNAL_BUILDER_METHODS = frozenset({
    "add_set", "add_delete", "get_operations", "is_empty",
    "get_capabilities", "mappers", "version", "_operations", "m",
})


# ============================================================================
# Section parse helpers
# ============================================================================

# Maps VyOS hyphenated sub-names to ConfigSyncSections field name suffixes
_IFACE_SUBTYPES: Dict[str, str] = {
    "bonding": "bonding",
    "bridge": "bridge",
    "dummy": "dummy",
    "ethernet": "ethernet",
    "geneve": "geneve",
    "input": "input",
    "l2tpv3": "l2tpv3",
    "loopback": "loopback",
    "macsec": "macsec",
    "openvpn": "openvpn",
    "pppoe": "pppoe",
    "pseudo-ethernet": "pseudo_ethernet",
    "sstpc": "sstpc",
    "tunnel": "tunnel",
    "virtual-ethernet": "virtual_ethernet",
    "vti": "vti",
    "vxlan": "vxlan",
    "wireguard": "wireguard",
    "wireless": "wireless",
    "wwan": "wwan",
}

_PROTO_SUBTYPES: Dict[str, str] = {
    "babel": "babel",
    "bfd": "bfd",
    "bgp": "bgp",
    "failover": "failover",
    "igmp-proxy": "igmp_proxy",
    "isis": "isis",
    "mpls": "mpls",
    "nhrp": "nhrp",
    "ospf": "ospf",
    "ospfv3": "ospfv3",
    "pim": "pim",
    "pim6": "pim6",
    "rip": "rip",
    "ripng": "ripng",
    "rpki": "rpki",
    "segment-routing": "segment_routing",
    "static": "static",
}

_QOS_SUBTYPES: Dict[str, str] = {
    "interface": "interface",
    "policy": "policy",
}

_SERVICE_SUBTYPES: Dict[str, str] = {
    "console-server": "console_server",
    "dhcp-relay": "dhcp_relay",
    "dhcp-server": "dhcp_server",
    "dhcpv6-relay": "dhcpv6_relay",
    "dhcpv6-server": "dhcpv6_server",
    "dns": "dns",
    "lldp": "lldp",
    "mdns": "mdns",
    "monitoring": "monitoring",
    "ndp-proxy": "ndp_proxy",
    "ntp": "ntp",
    "snmp": "snmp",
    "tftp-server": "tftp_server",
    "webproxy": "webproxy",
}

_SYSTEM_SUBTYPES: Dict[str, str] = {
    "conntrack": "conntrack",
    "flow-accounting": "flow_accounting",
    "login": "login",
    "option": "option",
    "sflow": "sflow",
    "static-host-mapping": "static_host_mapping",
    "sysctl": "sysctl",
    "time-zone": "time_zone",
}


def _parse_sections(section_raw: dict) -> ConfigSyncSections:
    """Parse the 'section' dict from VyOS config into a ConfigSyncSections model."""
    s = ConfigSyncSections()
    if not isinstance(section_raw, dict):
        return s

    def _has(key: str) -> bool:
        return key in section_raw

    def _children(key: str) -> dict:
        val = section_raw.get(key)
        return val if isinstance(val, dict) else {}

    s.firewall = _has("firewall")
    s.nat = _has("nat")
    s.nat66 = _has("nat66")
    s.pki = _has("pki")
    s.policy = _has("policy")
    s.vpn = _has("vpn")
    s.vrf = _has("vrf")

    if _has("interfaces"):
        s.interfaces = True
        iface_children = _children("interfaces")
        for vyos_name, field_suffix in _IFACE_SUBTYPES.items():
            if vyos_name in iface_children:
                setattr(s, f"interfaces_{field_suffix}", True)

    if _has("protocols"):
        s.protocols = True
        proto_children = _children("protocols")
        for vyos_name, field_suffix in _PROTO_SUBTYPES.items():
            if vyos_name in proto_children:
                setattr(s, f"protocols_{field_suffix}", True)

    if _has("qos"):
        s.qos = True
        qos_children = _children("qos")
        for vyos_name, field_suffix in _QOS_SUBTYPES.items():
            if vyos_name in qos_children:
                setattr(s, f"qos_{field_suffix}", True)

    if _has("service"):
        s.service = True
        svc_children = _children("service")
        for vyos_name, field_suffix in _SERVICE_SUBTYPES.items():
            if vyos_name in svc_children:
                setattr(s, f"service_{field_suffix}", True)

    if _has("system"):
        s.system = True
        sys_children = _children("system")
        for vyos_name, field_suffix in _SYSTEM_SUBTYPES.items():
            if vyos_name in sys_children:
                setattr(s, f"system_{field_suffix}", True)

    return s


# ============================================================================
# Endpoint 1: Capabilities
# ============================================================================


@router.get("/capabilities")
async def get_config_sync_capabilities(request: Request):
    """Return config-sync feature capabilities based on the VyOS version."""
    await require_read_permission(request, FeatureGroup.CONFIG_SYNC)
    try:
        service = get_session_vyos_service(request)
        builder = ConfigSyncBatchBuilder(version=service.get_version())
        caps = builder.get_capabilities()
        if hasattr(request.state, "instance") and request.state.instance:
            caps["instance_name"] = request.state.instance.get("name")
            caps["instance_id"] = request.state.instance.get("id")
        return caps
    except Exception:
        logger.exception("Unhandled error in get_config_sync_capabilities")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 2: Config
# ============================================================================


@router.get("/config", response_model=ConfigSyncConfig)
async def get_config_sync_config(http_request: Request, refresh: bool = False):
    """Return the full config-sync configuration in a normalized format."""
    await require_read_permission(http_request, FeatureGroup.CONFIG_SYNC)
    try:
        service = get_session_vyos_service(http_request)
        full_config = await run_in_threadpool(service.get_full_config, refresh=refresh)

        cs_raw = full_config.get("service", {}).get("config-sync", {})
        if not cs_raw:
            return ConfigSyncConfig()

        mode = cs_raw.get("mode")

        secondary = None
        sec_raw = cs_raw.get("secondary")
        if sec_raw and isinstance(sec_raw, dict):
            port_raw = sec_raw.get("port")
            timeout_raw = sec_raw.get("timeout")
            secondary = ConfigSyncSecondary(
                address=sec_raw.get("address"),
                key=sec_raw.get("key"),
                port=int(port_raw) if port_raw is not None else None,
                timeout=int(timeout_raw) if timeout_raw is not None else None,
            )

        sections = _parse_sections(cs_raw.get("section", {}))

        return ConfigSyncConfig(mode=mode, secondary=secondary, sections=sections)
    except Exception:
        logger.exception("Unhandled error in get_config_sync_config")
        raise HTTPException(status_code=500, detail="Internal server error")


# ============================================================================
# Endpoint 3: Batch Operations
# ============================================================================


@router.post("/batch", response_model=VyOSResponse)
async def config_sync_batch_configure(
    http_request: Request, body: ConfigSyncBatchRequest
):
    """Execute a batch of config-sync configuration operations atomically."""
    await require_write_permission(http_request, FeatureGroup.CONFIG_SYNC)
    try:
        service = get_session_vyos_service(http_request)
        builder = ConfigSyncBatchBuilder(version=service.get_version())

        for operation in body.operations:
            if operation.op in _INTERNAL_BUILDER_METHODS or operation.op.startswith("_"):
                raise HTTPException(
                    status_code=400,
                    detail=f"Operation not allowed: {operation.op}",
                )

            method = getattr(builder, operation.op)
            sig = inspect.signature(method)
            params = [p for p in sig.parameters.keys() if p != "self"]

            if len(params) == 0:
                method()
            elif len(params) == 1:
                if operation.value is not None:
                    method(operation.value)
                else:
                    method()
            elif len(params) >= 2:
                if operation.value and "," in operation.value:
                    parts = operation.value.split(",", len(params) - 1)
                    method(*parts)
                elif operation.value:
                    method(operation.value)

        response = service.execute_batch(builder)
        return VyOSResponse(
            success=response.status == 200,
            data={"message": "Config-Sync configuration updated"},
            error=response.error if response.error else None,
        )
    except HTTPException:
        raise
    except AttributeError as e:
        raise HTTPException(status_code=400, detail=f"Unknown operation: {e}")
    except Exception:
        logger.exception("Unhandled error in config_sync_batch_configure")
        raise HTTPException(status_code=500, detail="Internal server error")

File: frontend/src/lib/api/config-sync.ts
import { apiClient } from "./client";

export interface ConfigSyncSecondary {
  address?: string | null;
  key?: string | null;
  port?: number | null;
  timeout?: number | null;
}

export interface ConfigSyncSections {
  firewall: boolean;
  nat: boolean;
  nat66: boolean;
  pki: boolean;
  policy: boolean;
  vpn: boolean;
  vrf: boolean;

  interfaces: boolean;
  interfaces_bonding: boolean;
  interfaces_bridge: boolean;
  interfaces_dummy: boolean;
  interfaces_ethernet: boolean;
  interfaces_geneve: boolean;
  interfaces_input: boolean;
  interfaces_l2tpv3: boolean;
  interfaces_loopback: boolean;
  interfaces_macsec: boolean;
  interfaces_openvpn: boolean;
  interfaces_pppoe: boolean;
  interfaces_pseudo_ethernet: boolean;
  interfaces_sstpc: boolean;
  interfaces_tunnel: boolean;
  interfaces_virtual_ethernet: boolean;
  interfaces_vti: boolean;
  interfaces_vxlan: boolean;
  interfaces_wireguard: boolean;
  interfaces_wireless: boolean;
  interfaces_wwan: boolean;

  protocols: boolean;
  protocols_babel: boolean;
  protocols_bfd: boolean;
  protocols_bgp: boolean;
  protocols_failover: boolean;
  protocols_igmp_proxy: boolean;
  protocols_isis: boolean;
  protocols_mpls: boolean;
  protocols_nhrp: boolean;
  protocols_ospf: boolean;
  protocols_ospfv3: boolean;
  protocols_pim: boolean;
  protocols_pim6: boolean;
  protocols_rip: boolean;
  protocols_ripng: boolean;
  protocols_rpki: boolean;
  protocols_segment_routing: boolean;
  protocols_static: boolean;

  qos: boolean;
  qos_interface: boolean;
  qos_policy: boolean;

  service: boolean;
  service_console_server: boolean;
  service_dhcp_relay: boolean;
  service_dhcp_server: boolean;
  service_dhcpv6_relay: boolean;
  service_dhcpv6_server: boolean;
  service_dns: boolean;
  service_lldp: boolean;
  service_mdns: boolean;
  service_monitoring: boolean;
  service_ndp_proxy: boolean;
  service_ntp: boolean;
  service_snmp: boolean;
  service_tftp_server: boolean;
  service_webproxy: boolean;

  system: boolean;
  system_conntrack: boolean;
  system_flow_accounting: boolean;
  system_login: boolean;
  system_option: boolean;
  system_sflow: boolean;
  system_static_host_mapping: boolean;
  system_sysctl: boolean;
  system_time_zone: boolean;
}

export interface ConfigSyncConfig {
  mode?: string | null;
  secondary?: ConfigSyncSecondary | null;
  sections: ConfigSyncSections;
}

export interface ConfigSyncCapabilities {
  version: string;
  features: Record<string, unknown>;
  version_info: { is_1_4: boolean; is_1_5: boolean };
}

export interface VyOSResponse {
  success: boolean;
  data?: Record<string, unknown> | null;
  error?: string | null;
}

export interface BatchOperation {
  op: string;
  value?: string | null;
}

export function defaultSections(): ConfigSyncSections {
  return {
    firewall: false, nat: false, nat66: false, pki: false, policy: false, vpn: false, vrf: false,
    interfaces: false, interfaces_bonding: false, interfaces_bridge: false, interfaces_dummy: false,
    interfaces_ethernet: false, interfaces_geneve: false, interfaces_input: false, interfaces_l2tpv3: false,
    interfaces_loopback: false, interfaces_macsec: false, interfaces_openvpn: false, interfaces_pppoe: false,
    interfaces_pseudo_ethernet: false, interfaces_sstpc: false, interfaces_tunnel: false,
    interfaces_virtual_ethernet: false, interfaces_vti: false, interfaces_vxlan: false,
    interfaces_wireguard: false, interfaces_wireless: false, interfaces_wwan: false,
    protocols: false, protocols_babel: false, protocols_bfd: false, protocols_bgp: false,
    protocols_failover: false, protocols_igmp_proxy: false, protocols_isis: false, protocols_mpls: false,
    protocols_nhrp: false, protocols_ospf: false, protocols_ospfv3: false, protocols_pim: false,
    protocols_pim6: false, protocols_rip: false, protocols_ripng: false, protocols_rpki: false,
    protocols_segment_routing: false, protocols_static: false,
    qos: false, qos_interface: false, qos_policy: false,
    service: false, service_console_server: false, service_dhcp_relay: false, service_dhcp_server: false,
    service_dhcpv6_relay: false, service_dhcpv6_server: false, service_dns: false, service_lldp: false,
    service_mdns: false, service_monitoring: false, service_ndp_proxy: false, service_ntp: false,
    service_snmp: false, service_tftp_server: false, service_webproxy: false,
    system: false, system_conntrack: false, system_flow_accounting: false, system_login: false,
    system_option: false, system_sflow: false, system_static_host_mapping: false, system_sysctl: false,
    system_time_zone: false,
  };
}

// Maps TypeScript field name → [parent, vyos-name] for sub-sections
const SUB_SECTION_MAP: Record<string, [string, string]> = {
  interfaces_bonding: ["interfaces", "bonding"],
  interfaces_bridge: ["interfaces", "bridge"],
  interfaces_dummy: ["interfaces", "dummy"],
  interfaces_ethernet: ["interfaces", "ethernet"],
  interfaces_geneve: ["interfaces", "geneve"],
  interfaces_input: ["interfaces", "input"],
  interfaces_l2tpv3: ["interfaces", "l2tpv3"],
  interfaces_loopback: ["interfaces", "loopback"],
  interfaces_macsec: ["interfaces", "macsec"],
  interfaces_openvpn: ["interfaces", "openvpn"],
  interfaces_pppoe: ["interfaces", "pppoe"],
  interfaces_pseudo_ethernet: ["interfaces", "pseudo-ethernet"],
  interfaces_sstpc: ["interfaces", "sstpc"],
  interfaces_tunnel: ["interfaces", "tunnel"],
  interfaces_virtual_ethernet: ["interfaces", "virtual-ethernet"],
  interfaces_vti: ["interfaces", "vti"],
  interfaces_vxlan: ["interfaces", "vxlan"],
  interfaces_wireguard: ["interfaces", "wireguard"],
  interfaces_wireless: ["interfaces", "wireless"],
  interfaces_wwan: ["interfaces", "wwan"],
  protocols_babel: ["protocols", "babel"],
  protocols_bfd: ["protocols", "bfd"],
  protocols_bgp: ["protocols", "bgp"],
  protocols_failover: ["protocols", "failover"],
  protocols_igmp_proxy: ["protocols", "igmp-proxy"],
  protocols_isis: ["protocols", "isis"],
  protocols_mpls: ["protocols", "mpls"],
  protocols_nhrp: ["protocols", "nhrp"],
  protocols_ospf: ["protocols", "ospf"],
  protocols_ospfv3: ["protocols", "ospfv3"],
  protocols_pim: ["protocols", "pim"],
  protocols_pim6: ["protocols", "pim6"],
  protocols_rip: ["protocols", "rip"],
  protocols_ripng: ["protocols", "ripng"],
  protocols_rpki: ["protocols", "rpki"],
  protocols_segment_routing: ["protocols", "segment-routing"],
  protocols_static: ["protocols", "static"],
  qos_interface: ["qos", "interface"],
  qos_policy: ["qos", "policy"],
  service_console_server: ["service", "console-server"],
  service_dhcp_relay: ["service", "dhcp-relay"],
  service_dhcp_server: ["service", "dhcp-server"],
  service_dhcpv6_relay: ["service", "dhcpv6-relay"],
  service_dhcpv6_server: ["service", "dhcpv6-server"],
  service_dns: ["service", "dns"],
  service_lldp: ["service", "lldp"],
  service_mdns: ["service", "mdns"],
  service_monitoring: ["service", "monitoring"],
  service_ndp_proxy: ["service", "ndp-proxy"],
  service_ntp: ["service", "ntp"],
  service_snmp: ["service", "snmp"],
  service_tftp_server: ["service", "tftp-server"],
  service_webproxy: ["service", "webproxy"],
  system_conntrack: ["system", "conntrack"],
  system_flow_accounting: ["system", "flow-accounting"],
  system_login: ["system", "login"],
  system_option: ["system", "option"],
  system_sflow: ["system", "sflow"],
  system_static_host_mapping: ["system", "static-host-mapping"],
  system_sysctl: ["system", "sysctl"],
  system_time_zone: ["system", "time-zone"],
};

const SIMPLE_SECTIONS = ["firewall", "nat", "nat66", "pki", "policy", "vpn", "vrf"] as const;
const PARENT_SECTIONS = ["interfaces", "protocols", "qos", "service", "system"] as const;

class ConfigSyncService {
  async getCapabilities(): Promise<ConfigSyncCapabilities> {
    return apiClient.get<ConfigSyncCapabilities>("/vyos/config-sync/capabilities");
  }

  async getConfig(refresh = false): Promise<ConfigSyncConfig> {
    return apiClient.get<ConfigSyncConfig>("/vyos/config-sync/config", {
      refresh: refresh.toString(),
    });
  }

  private async batch(operations: BatchOperation[]): Promise<VyOSResponse> {
    const result = await apiClient.post<VyOSResponse>("/vyos/config-sync/batch", { operations });
    if (!result.success) {
      throw new Error(result.error || "Operation failed");
    }
    return result;
  }

  async saveConfig(original: ConfigSyncConfig | null, updated: ConfigSyncConfig): Promise<VyOSResponse> {
    const ops: BatchOperation[] = [];
    const orig = original ?? { mode: null, secondary: null, sections: defaultSections() };

    // Mode
    const oldMode = orig.mode ?? null;
    const newMode = updated.mode ?? null;
    if (oldMode !== newMode) {
      if (newMode) {
        ops.push({ op: "set_mode", value: newMode });
      } else {
        ops.push({ op: "delete_mode" });
      }
    }

    // Secondary fields
    const oldSec = orig.secondary ?? {};
    const newSec = updated.secondary ?? {};

    const secFields: Array<{ key: keyof ConfigSyncSecondary; setOp: string; delOp: string }> = [
      { key: "address", setOp: "set_secondary_address", delOp: "delete_secondary_address" },
      { key: "key", setOp: "set_secondary_key", delOp: "delete_secondary_key" },
      { key: "port", setOp: "set_secondary_port", delOp: "delete_secondary_port" },
      { key: "timeout", setOp: "set_secondary_timeout", delOp: "delete_secondary_timeout" },
    ];

    for (const { key, setOp, delOp } of secFields) {
      const oldVal = (oldSec as Record<string, unknown>)[key] ?? null;
      const newVal = (newSec as Record<string, unknown>)[key] ?? null;
      const oldStr = oldVal != null ? String(oldVal) : null;
      const newStr = newVal != null && String(newVal) !== "" ? String(newVal) : null;
      if (oldStr !== newStr) {
        if (newStr) {
          ops.push({ op: setOp, value: newStr });
        } else if (oldStr) {
          ops.push({ op: delOp });
        }
      }
    }

    // Simple sections
    const oldSections = orig.sections ?? defaultSections();
    const newSections = updated.sections ?? defaultSections();

    for (const name of SIMPLE_SECTIONS) {
      const wasOn = (oldSections as unknown as Record<string, boolean>)[name] ?? false;
      const isOn = (newSections as unknown as Record<string, boolean>)[name] ?? false;
      if (wasOn !== isOn) {
        ops.push({ op: isOn ? "set_section" : "delete_section", value: name });
      }
    }

    // Parent sections
    for (const name of PARENT_SECTIONS) {
      const wasOn = (oldSections as unknown as Record<string, boolean>)[name] ?? false;
      const isOn = (newSections as unknown as Record<string, boolean>)[name] ?? false;
      if (wasOn !== isOn) {
        ops.push({ op: isOn ? "set_section" : "delete_section", value: name });
      }
    }

    // Sub-sections
    for (const [field, [parent, vyosName]] of Object.entries(SUB_SECTION_MAP)) {
      const wasOn = (oldSections as unknown as Record<string, boolean>)[field] ?? false;
      const isOn = (newSections as unknown as Record<string, boolean>)[field] ?? false;
      if (wasOn !== isOn) {
        ops.push({ op: isOn ? "set_section_sub" : "delete_section_sub", value: `${parent},${vyosName}` });
      }
    }

    if (ops.length === 0) {
      return { success: true, data: { message: "No changes" } };
    }

    return this.batch(ops);
  }

  async deleteConfigSync(): Promise<VyOSResponse> {
    return this.batch([{ op: "delete_config_sync" }]);
  }
}

export const configSyncService = new ConfigSyncService();
